UEFI SB release

tomzy-0 opened the pull request:
This Pull Request gather changes made in OSFV after working on Secure Boot Test Suite for MPL PIP platform that uses AMI BIOS. Following modifications are introduced here:

* Added new UEFI Secure Boot test cases: SBO009 - SBO015
* Split BIOS menu related keywords into Dasharo/AMI/common
  - changes in multiple files to drop `Tianocore` from names
  - KW names duplicates between Dasharo/AMI so the test case can use the same KW names but depending on used platform configuration, Dasharo or AMI definition of given KW will be used 
* Add script to generate images used for UEFI Secure Boot tests
* Introduce ansible usage as a part of Suite Setup, to perform additional actions on booted OSes
  - adds readme entry about usage
* Added couple new variables to platform-configs

**What could be changed**

* Some tests depends on booting DTS with UEFI SB integrated, need a way how to send its image if it is not on PiKVM
* Ansible command and hosts configuration could be handled better
  - now for every new device we need to add new hosts entry
  - ansible is run for every host form hosts file, failures on not available hosts are skipped

Commit:
.gitignore: add qemu-data used in tests

Signed-off-by: Tomasz Żyjewski <tomasz.zyjewski@3mdeb.com>

Commit:
scripts: qemu-run: fix typo in warning comment

Signed-off-by: Tomasz Żyjewski <tomasz.zyjewski@3mdeb.com>

Commit:
lib: QemuMonitor: set ROBOT_LIBRARY_SCOPE to SUITE

To use Power On keyword in Setup Suite section, we need to set
ROBOT_LIBRARY_SCOPE variable.
See: http://robotframework.org/robotframework/latest/RobotFrameworkUserGuide.html#test-library-scope

This is necessary to use KW from this lib in setup that runs
ansible-playbooks on target.

Signed-off-by: Tomasz Żyjewski <tomasz.zyjewski@3mdeb.com>

Commit:
scripts: qemu-run: Add DRIVE_PATH, which allows adding virtual drives

Signed-off-by: Tomasz Żyjewski <tomasz.zyjewski@3mdeb.com>

Commit:
platform-configs: qemu: add variable used in tests with DTS and UEFI SB enabled

see: https://github.com/Wind-River/meta-secure-core/tree/master/meta-efi-secure-boot#quick-start-for-the-first-boot

This is a procedure integrated in DTS for UEFI Secure Boot enabled
build.

see: https://github.com/Dasharo/meta-dts/blob/main/kas-uefi-sb.yml

Signed-off-by: Tomasz Żyjewski <tomasz.zyjewski@3mdeb.com>

Commit:
lib: Split menus and secure-boot into ami/dasharo files

Signed-off-by: Tomasz Żyjewski <tomasz.zyjewski@3mdeb.com>

Commit:
platform-configs: resource needed bios and secure-boot robot files

For those platforms we are using Dasharo firmware

Signed-off-by: Tomasz Żyjewski <tomasz.zyjewski@3mdeb.com>

Commit:
variables: add MPL-PIP4 platform

Signed-off-by: Tomasz Żyjewski <tomasz.zyjewski@3mdeb.com>

Commit:
platform-configs: mpl-pip4: add configuration for MPL-PIP4 device

Signed-off-by: Tomasz Żyjewski <tomasz.zyjewski@3mdeb.com>

m-iwanicki committed:
lib: dts-lib: shorten pikvm DTS name due to AMI limit

Don't wait for DTS to boot when certificate_provisioning == True
when using pikvm connection method

Signed-off-by: Michał Iwanicki <michal.iwanicki@3mdeb.com>

Commit:
lib: ansible: add KW to restore connection method

Needed as sometimes RF change connection to SSH

Signed-off-by: Tomasz Żyjewski <tomasz.zyjewski@3mdeb.com>

Commit:
variables: use IP address for MPL sonoff

Signed-off-by: Tomasz Żyjewski <tomasz.zyjewski@3mdeb.com>

m-iwanicki committed:
lib: secure-boot-dasharo: change default value

of sb_menu in Enter Key Management And... to empty list instead of
empty string.

Signed-off-by: Michał Iwanicki <michal.iwanicki@3mdeb.com>

Commit:
lib: bios: menus-dasharo: fix typo for checkpoint

Signed-off-by: Tomasz Żyjewski <tomasz.zyjewski@3mdeb.com>

Commit:
dasharo-security: secure-boot: add SBO009.001

Signed-off-by: Tomasz Żyjewski <tomasz.zyjewski@3mdeb.com>

Commit:
dasharo-security: secure-boot: add SBO010.001-006

Signed-off-by: Tomasz Żyjewski <tomasz.zyjewski@3mdeb.com>

Commit:
dasharo-security: secure-boot: add SBO011.001

Signed-off-by: Tomasz Żyjewski <tomasz.zyjewski@3mdeb.com>

Commit:
dasharo-security: secure-boot: add SBO012.001

Add needed keywords in secure-boot-lib and fix skips in dasharo-security

Signed-off-by: Tomasz Żyjewski <tomasz.zyjewski@3mdeb.com>

Commit:
Introduce ansible usage in osfv

* adds new packages to requirements.txt
* adds ansible-roles directory with tests example of usage
* set default value for ANSIBLE_CONFIG variable
* adds section in README about ansible usage
* adds Run Ansible Playbook On Supported Operating Systems keyword
* test ansible runs on secure-boot.robot

Signed-off-by: Tomasz Żyjewski <tomasz.zyjewski@3mdeb.com>

Commit:
lib: secure-boot-lib: add keywords for UEFI SB tests in OS

Signed-off-by: Tomasz Żyjewski <tomasz.zyjewski@3mdeb.com>

Commit:
dasharo-security: secure-boot: add SBO013.001-002 and SBO014.001

Signed-off-by: Tomasz Żyjewski <tomasz.zyjewski@3mdeb.com>

Commit:
dasharo-security: secure-boot: add SBO015.001

Expand Enroll Secure Boot Keys In OS KW, add fixes to SBO014.001
and SBO012.001

Signed-off-by: Tomasz Żyjewski <tomasz.zyjewski@3mdeb.com>

Commit:
Introduce sb-img-wrapper script

The script generates ISO images with efi executables used in UEFI Secure
Boot tests (where we test different scenarios of good/bad signed files
or with differnet cryptographic algorithms).

Provides new keyword that allows to mount ISO images as USB in BIOS
Menu.

Signed-off-by: Tomasz Żyjewski <tomasz.zyjewski@3mdeb.com>

Commit:
Use common BIOS menu keywords

Rename used keyword from Tianocore based to common using
scripts/rename-keyword.sh script.

Signed-off-by: Tomasz Żyjewski <tomasz.zyjewski@3mdeb.com>

Commit:
keys: add PAGEUP/DOWN, HOME, END as Global Variables

Signed-off-by: Tomasz Żyjewski <tomasz.zyjewski@3mdeb.com>

Commit:
Add ansible configuration for MPL PIP device

Every new platform needs new entry in ansible-roles/hosts file

Signed-off-by: Tomasz Żyjewski <tomasz.zyjewski@3mdeb.com>

Commit:
dasharo-security: secure-boot: reimplement tests after divided to ami/dasharo

Signed-off-by: Tomasz Żyjewski <tomasz.zyjewski@3mdeb.com>

Commit:
lib: move more keywords to common files after tests on MPL

Signed-off-by: Tomasz Żyjewski <tomasz.zyjewski@3mdeb.com>

Commit:
lib: ansible: add argument to ansible run KW

Provide dir for ansible runs per test suite and test on
secure-boot.robot. Remove ubuntu example. Modify lib/ansible.robot
to use argument to choose different playbook based on the test suite
name.

Signed-off-by: Tomasz Żyjewski <tomasz.zyjewski@3mdeb.com>

Commit:
platform-config: introduce SECURE_BOOT_CAN_REMOVE_EXTERNAL_CERT variable

With AMI BIOS we are not able to reset certs enrolled from OS, to handle
such cases we added new flag to skip some tests

Signed-off-by: Tomasz Żyjewski <tomasz.zyjewski@3mdeb.com>

Commit:
dasharo-security: secure-boot: fix tests after execution on AMI

Signed-off-by: Tomasz Żyjewski <tomasz.zyjewski@3mdeb.com>

Commit:
variables: fixup RTE_LIST variable

Signed-off-by: Tomasz Żyjewski <tomasz.zyjewski@3mdeb.com>

Commit:
platform-configs: mpl-pip4: run robotidy

Signed-off-by: Tomasz Żyjewski <tomasz.zyjewski@3mdeb.com>

Commit:
dasharo-security: secure-boot: update tests after tests on QEMU

Signed-off-by: Tomasz Żyjewski <tomasz.zyjewski@3mdeb.com>

Commit:
lib: dts-lib: fixes after tests on QEMU

Signed-off-by: Tomasz Żyjewski <tomasz.zyjewski@3mdeb.com>

Commit:
lib: bios: move Exit From Current Menu to common

Signed-off-by: Tomasz Żyjewski <tomasz.zyjewski@3mdeb.com>

Commit:
self-tests: rename leftovers after droping Tianocore from KW names

Signed-off-by: Tomasz Żyjewski <tomasz.zyjewski@3mdeb.com>

TomaszAIR commented:
Results on QEMU after latest rebase on `develop` below. QEMU started with `DRIVE_PATH=./qemu-data/dts-base-sb-image-genericx86-64.wic ./scripts/ci/qemu-run.sh graphic os` command, `dts-base-sb-image-genericx86-64.wic` is decompressed DTS image built by using https://github.com/Dasharo/meta-dts/blob/main/kas-uefi-sb.yml.

```
λ robot -b file.txt -L TRACE -v ansible_config:yes -v rte_ip:127.0.0.1 -v snipeit:no -v config:qemu dasharo-security/secure-boot.robot
==============================================================================
Secure-Boot
==============================================================================
SBO001.001 Check Secure Boot default state (firmware) :: This test... | PASS |
------------------------------------------------------------------------------
SBO002.001 UEFI Secure Boot (Ubuntu 22.04) :: This test verifies t... | PASS |
------------------------------------------------------------------------------
SBO002.002 UEFI Secure Boot (Windows 11) :: This test verifies tha... | SKIP |
SBO002.002 not supported
------------------------------------------------------------------------------
SBO003.001 Attempt to boot file with the correct key from Shell (f... | PASS |
------------------------------------------------------------------------------
SBO004.001 Attempt to boot file without the key from Shell (firmwa... | PASS |
------------------------------------------------------------------------------
SBO005.001 Attempt to boot file with the wrong-signed key from She... | PASS |
------------------------------------------------------------------------------
SBO006.001 Reset Secure Boot Keys option availability (firmware) :... | PASS |
------------------------------------------------------------------------------
SBO007.001 Attempt to boot the file after restoring keys to defaul... | PASS |
------------------------------------------------------------------------------
SBO008.001 Attempt to enroll the key in the incorrect format (firm... | PASS |
------------------------------------------------------------------------------
SBO009.001 Attempt to boot file signed for intermediate certificat... | PASS |
------------------------------------------------------------------------------
SBO010.001 Check support for rsa2k signed certificates :: PEM gene... | PASS |
------------------------------------------------------------------------------
SBO010.002 Check support for rsa3k signed certificates :: PEM gene... | PASS |
------------------------------------------------------------------------------
SBO010.003 Check support for rsa4k signed certificates :: PEM gene... | PASS |
------------------------------------------------------------------------------
SBO010.004 Check support for ecdsa256 signed certificates :: PEM g... | FAIL |
'
Command Error Status: Access Denied
FS0:\>' does not contain 'hello, world!'
------------------------------------------------------------------------------
SBO010.005 Check support for ecdsa384 signed certificates :: PEM g... | FAIL |
'
Command Error Status: Access Denied
FS0:\>' does not contain 'hello, world!'
------------------------------------------------------------------------------
SBO010.006 Check support for ecdsa521 signed certificates :: PEM g... | FAIL |
'
Command Error Status: Access Denied
FS0:\>' does not contain 'hello, world!'
------------------------------------------------------------------------------
SBO011.001 Attempt to enroll expired certificate and boot signed i... | FAIL |
'
Hello, world!
FS0:\>' does not contain any of 'Access Denied'
------------------------------------------------------------------------------
SBO012.001 Boot OS Signed And Enrolled From Inside System (Ubuntu ... | PASS |
------------------------------------------------------------------------------
SBO013.001 Check automatic certificate provisioning :: This test v... | PASS |
------------------------------------------------------------------------------
SBO013.002 Check automatic certificate provisioning KEK certificat... | PASS |
------------------------------------------------------------------------------
SBO014.001 Enroll certificates using sbctl :: This test erases Sec... | PASS |
------------------------------------------------------------------------------
SBO015.001 Attempt to enroll the key in the incorrect format (OS) ... | PASS |
------------------------------------------------------------------------------
Secure-Boot                                                           | FAIL |
22 tests, 17 passed, 4 failed, 1 skipped
==============================================================================
Debug:   /home/tzyjewski/files/open-source-firmware-validation/file.txt
Output:  /home/tzyjewski/files/open-source-firmware-validation/output.xml
Log:     /home/tzyjewski/files/open-source-firmware-validation/log.html
Report:  /home/tzyjewski/files/open-source-firmware-validation/report.html
```

ecdsa tests fails as BIOS does not accepts ecdsa-based certs

test with expired cert fails, as BIOS can still execute efi file signed with expired cert

Commit:
platform-configs: propagate ANSIBLE_SUPPORT flag

Signed-off-by: Tomasz Żyjewski <tomasz.zyjewski@3mdeb.com>

Commit:
ansible-roles: hosts: add host for msi-pro-z690-ddr5 ubuntu

Signed-off-by: Tomasz Żyjewski <tomasz.zyjewski@3mdeb.com>

Commit:
platform-configs: add missing flags

DTS_UEFI_SB_SUPPORT is used in SBO013 tests and needs DTS with UEFI
Secure Boot functionality

SECURE_BOOT_CAN_REMOVE_EXTERNAL_CERT is used when certificates enrolled
from OS can be removed on the BIOS menu level

Signed-off-by: Tomasz Żyjewski <tomasz.zyjewski@3mdeb.com>

Commit:
lib: usb-hid-msc-lib: update docs on Download ISO And Mount As USB

Signed-off-by: Tomasz Żyjewski <tomasz.zyjewski@3mdeb.com>

Commit:
dasharo-security: secure-boot: mount dts-sb.wic image before 013 tests

Mount dts-sb.wic which is an image build with
https://github.com/Dasharo/meta-dts/blob/main/kas-uefi-sb.yml

When testing with QEMU, dts-sb.wic should be placed in
./qemu-data/dts-sb.wic

Signed-off-by: Tomasz Żyjewski <tomasz.zyjewski@3mdeb.com>

Commit:
keywords: use Global instead of Test variables

Setting to Test was introduced in
https://github.com/Dasharo/open-source-firmware-validation/commit/5ea2291bd5f7c,
but some of those KW (e.g. Login To Linux) are used in Suite Setup, and
because of that, the Setup fails
see: https://robotframework.org/robotframework/latest/libraries/BuiltIn.html#Set%20Test%20Variable

Signed-off-by: Tomasz Żyjewski <tomasz.zyjewski@3mdeb.com>

Commit:
scripts: secure-boot: sb-img-wrapper: add usage and other functions

Signed-off-by: Tomasz Żyjewski <tomasz.zyjewski@3mdeb.com>

Commit:
Get rid of Tianocore from names in common KWs

Signed-off-by: Tomasz Żyjewski <tomasz.zyjewski@3mdeb.com>

Commit:
lib: ansible: remove TODOs

First TODO is not yet tested as we only used Ubuntu, next could be
Windows
see: https://docs.ansible.com/ansible/latest/os_guide/windows_usage.html

Second TODO was resolved by testing ansible config on MSI Pro Z690-A
DDR5.

For third TODO we have an info in README about how new hosts need to be
added, before running ansible config in tests.

Signed-off-by: Tomasz Żyjewski <tomasz.zyjewski@3mdeb.com>

Commit:
platform-configs: mpl-pip4: remove unused variables, resolve precommit

Signed-off-by: Tomasz Żyjewski <tomasz.zyjewski@3mdeb.com>

TomaszAIR commented:
Latest results on msi pro z690-a ddr5 
[results.tar.gz](https://github.com/Dasharo/open-source-firmware-validation/files/15091422/results.tar.gz)

Sometimes suite teardown fails (or any other test) because test e.g. goes into `HDD Security Configuration` instead of `Secure Boot Configuration`. Looks like it sometimes slip one step.
[results-teardown.tar.gz](https://github.com/Dasharo/open-source-firmware-validation/files/15091453/results-teardown.tar.gz)

macpijan commented on lib/secure-boot-lib-common.robot:
```diff
@@ -0,0 +1,122 @@
+    [Documentation]    Removes all files and directories in
+    ...    `/usr/share/secureboot` which is used by sbctl in the operating
+    ...    system
+    Execute Linux Command Without Output    rm -rf /usr/share/secureboot
```
Please use `Terminal` keywords instead.
Also, I assume we should not use the `without output` version, as we should wait for the next prompt, to make sure the removal has finished.

TomaszAIR replied:
Added but this was not tested yet https://github.com/Dasharo/open-source-firmware-validation/pull/187/commits/1b85aa7b649fa94a881a79f22f2f646bd04d18d4

macpijan commented on lib/secure-boot-lib-common.robot:
```diff
@@ -0,0 +1,122 @@
+Generate Secure Boot Keys In OS
+    [Documentation]    Generates new Secure Boot keys using sbctl.
+    ...    Returns true if succeeded.
+    ${out}=    Execute Linux Command    sbctl create-keys
```
Use Terminal kwds. In the whole MR.

TomaszAIR replied:
Added but this was not tested yet https://github.com/Dasharo/open-source-firmware-validation/pull/187/commits/1b85aa7b649fa94a881a79f22f2f646bd04d18d4

I think that whole MR was checked and this KWD is not used in it.

macpijan commented on lib/secure-boot-lib-dasharo.robot:
```diff
@@ -107,6 +107,7 @@ Make Sure That Keys Are Provisioned
     IF    ${need_reset} == ${TRUE}
         ${advanced_menu}=    Enter Advanced Secure Boot Keys Management And Return Construction    ${sb_menu}
         Reset To Default Secure Boot Keys    ${advanced_menu}
+        Sleep    3s
```
Any comment on this?

m-iwanicki replied:
@macpijan Related to https://github.com/Dasharo/open-source-firmware-validation/issues/154

TomaszAIR replied:
Probably tests failed here when running `Exit From Current Menu` KWD after `Press Enter` from `Reset To Default Secure Boot Keys`. I would say that we did not found better solution to make sure that exiting can be run.

macpijan commented on lib/usb-hid-msc-lib.robot:
```diff
@@ -33,3 +33,23 @@ Download ISO And Mount As USB
+
+Mount ISO As USB
+    [Documentation]    Mounts an ISO file from the local sources. In case of
+    ...    PiKVM we need to manually upload images to the device. For UEFI
```
We don't, we can upload images via API as well, right? See: `Upload Image To PiKVM` kwd.

TomaszAIR replied:
Similar to https://github.com/Dasharo/open-source-firmware-validation/pull/187#discussion_r1601016146, I think we have only KWDs that needs url, so images generated with `scripts/secure-boot/generate-images/sb-img-wrapper.sh` would need to be

a) sent to cloud to have some link
b) hosted locally (?) so then we could start simple http server and upload from there

I think that we did not want to store those images anywhere (so the script to prepare images locally was created) and hosting them seems problematic? Wouldn't we need to know an ip address of the host?

macpijan commented on lib/usb-hid-msc-lib.robot:
```diff
@@ -33,3 +33,23 @@ Download ISO And Mount As USB
+            Mount Image On PiKVM    ${PIKVM_IP}    ${img_name}
+        ELSE
+            Skip    unsupported
+        END
```
Any informative error message on other platform? Or we already ensure this would not be called in such a case?

TomaszAIR replied:
Tried to add such msg https://github.com/Dasharo/open-source-firmware-validation/pull/187/commits/13e2018fef5b3b1f047a06eb4538705a8f4c1141 but to be honest, I think that whole IF with pikvm should be gone, because for real hardware we always use it

macpijan commented on scripts/ci/qemu-run.sh:
```diff
@@ -52,11 +52,15 @@ This is the QEMU wrapper script for the Dasharo Open Source Firmware Validation.
     ./$(basename $0) vnc firmware
     ./$(basename $0) graphic os_install
     DIR=/my/work/dir ./$(basename $0) graphic os
+    DRIVE_PATH=/path/to/image ./$(basename $0) graphic os
```
How it should be different from HDD_PATH?

TomaszAIR replied:
This was added to boot qemu with two images, Ubuntu installed on HDD_PATH and DTS added via DRIVE_PATH. @PLangowski I think you worked on that. Maybe we could use here some different logic to be able to mount multiple images?

macpijan commented on scripts/secure-boot/generate-images/generate-image.sh:
```diff
@@ -0,0 +1,216 @@
+#!/bin/bash
```
It seems we would be fine with just: `scripts/secure-boot` path

TomaszAIR replied:
Fixed https://github.com/Dasharo/open-source-firmware-validation/pull/187/commits/13e2018fef5b3b1f047a06eb4538705a8f4c1141

macpijan commented on variables.robot:
```diff
@@ -202,6 +203,12 @@ ${OS_UBUNTU}=               ubuntu
+# MPL-PIP4
+&{RTE56}=                   ip=192.168.4.165    cpuid=02c00042ea42e744    pcb_rev=1.0.0
+...                         platform=mpl-pip4    sonoff_ip=192.168.4.222
+...                         board-revision=x    env=dev    platform_vendor=mpl
```
Is `env` variables used?
Also others like: `cpuid` `board-revision` ?

TomaszAIR replied:
No, they are not. I think we used other RTE entries as template without much thinking about this. Fixed https://github.com/Dasharo/open-source-firmware-validation/pull/187/commits/29383be17cb72e36e5af37ff8c75c6921cc4ec66

macpijan commented on variables.robot:
```diff
@@ -337,7 +345,7 @@ ${OS_UBUNTU}=               ubuntu
 ...                         sbo_name=USB    name=SanDisk Ultra USB 3.0
 &{USB13}=                   vendor=Artificial    volume=1GB    type=USB_Storage
 ...                         protocol=3.0    interface=USB    count=1
-...                         sbo_name=USB    name=PiKVM Composite KVM Device
```
Why do we drop `Device` ?

TomaszAIR replied:
Limitation on AMI https://github.com/Dasharo/open-source-firmware-validation/commit/6685ced3689cfb524fb687d14855f4e557e12c11 We couldn't add whole name there

macpijan reviewed (commented):
@TomaszAIR left some comments after revieving

TomaszAIR commented on README.md:
```diff
@@ -114,6 +115,12 @@ pip install -r requirements.txt
+* Before running the tests from `dasharo-security/secure-boot.robot`, please run
+  the [sb-img-wrapper.sh](./scripts/secure-boot/generate-images/sb-img-wrapper.sh)
+  script. Its task is to generate ISO images with the certificates and efi files
+  used during tests. If you are running tests with PiKVM, you need to add PiKVM
```
I think we only have KWD for upload from URL https://github.com/Dasharo/open-source-firmware-validation/blob/uefi-sb-release/lib/usb-hid-msc-lib.robot#L13 and those images are generated locally. I do not recall any other KWD that could be used here so in test suite we use https://github.com/Dasharo/open-source-firmware-validation/blob/uefi-sb-release/lib/usb-hid-msc-lib.robot#L37 to mount ISO which is already on PIKVM thanks to this script.

TomaszAIR commented on README.md:
```diff
@@ -161,6 +168,9 @@ Parameters should be defined as follows:
 * $TEST_CASE_ID - ID of the requested to run test case (i.e. `CBP001.001*`).
   Note that after test case ID asterisk should be added, if you do not wish
   to provide the full test name here.
+* $ANSIBLE_CONFIG - if set to yes can run `ansible-playbook` on target to
```
IIRC we set default in variables.robot to `no` https://github.com/Dasharo/open-source-firmware-validation/blob/uefi-sb-release/variables.robot#L11

TomaszAIR commented on README.md:
```diff
@@ -184,6 +194,43 @@ platform config file.
+* use of the [Run Ansible Playbook On Supported Operating
+  Systems](./lib/ansible.robot) keyword in the Suite Setup section of the
+  running test suite,
+* preparation of the IP address and port information used to connect SSH to the
```
I know, but then I did not think of a way to set this runtime while running robot command so we hardcoded it in https://github.com/Dasharo/open-source-firmware-validation/blob/uefi-sb-release/ansible-roles/hosts#L2. Maybe it could be set while running `ansible-playbook` in https://github.com/Dasharo/open-source-firmware-validation/blob/uefi-sb-release/lib/ansible.robot#L24

TomaszAIR commented on README.md:
```diff
@@ -184,6 +194,43 @@ platform config file.
+  [ansible-roles/roles](./ansible-roles/roles) to be executed before starting the
+  tests, the idea here is to store playbooks yml files under
+  `${suite_test_name}/tasks/common.yml`,
+* set `ANSIBLE_SUPPORT` to `${TRUE}` in platform config file.
```
You are right, it should not be platform specific, rather OS specific. So maybe it should go to https://github.com/Dasharo/open-source-firmware-validation/blob/uefi-sb-release/os-config/ubuntu-credentials.robot if we would change its name to `ubuntu.robot` or smth like that.

I think the logic here was to add `ANSIBLE_SUPPORT` on the platforms we tested this.

TomaszAIR commented on ansible-roles/hosts:
```diff
@@ -0,0 +1,8 @@
+127.0.0.1:5222 ansible_user=ubuntu ansible_ssh_pass=ubuntu
+
+[mpl-pip4_ubuntu]
+192.168.4.186:22 ansible_user=ubuntu ansible_ssh_pass=ubuntu
```
Mentioned here https://github.com/Dasharo/open-source-firmware-validation/pull/187#discussion_r1601019902

Looking at docs, it should be possible to list ip address as `-i` flag https://docs.ansible.com/ansible/latest/cli/ansible-playbook.html#cmdoption-ansible-playbook-i and rest info (port, ssh_pass, user) as https://docs.ansible.com/ansible/latest/cli/ansible-playbook.html#cmdoption-ansible-playbook-ssh-common-args

Commit:
Change dir name from ansible-roles to ansible

Signed-off-by: Tomasz Żyjewski <tomasz.zyjewski@3mdeb.com>

TomaszAIR commented on README.md:
```diff
@@ -184,6 +194,43 @@ platform config file.
+  DUT in the [ansible-roles/hosts](./ansible-roles/hosts) file, along with
+  credentials for logging into the system,
+* preparation of the relevant playbook under
+  [ansible-roles/roles](./ansible-roles/roles) to be executed before starting the
```
https://github.com/Dasharo/open-source-firmware-validation/pull/187/commits/61840f2cd141fc81a70cf4cff921f89de760f59e

TomaszAIR commented on ansible-roles/roles/secure-boot/tasks/common.yml:
```diff
@@ -0,0 +1,37 @@
+---
+- name: Install dependency packages
```
While developing we named it `common` as Ubuntu was only OS we tested this on. It will not run on `Windows` because of https://github.com/Dasharo/open-source-firmware-validation/blob/uefi-sb-release/lib/ansible.robot#L16.

We could name it after an OS but then I think we would need some env var to say on which OS given test setup is ran. I did not want to add another variable set on cmdline line but also did not have another idea for this.

If there was a variable for this, to simplify whole ansible KWD we would also need to use that variable with `Boot System Or From Connected Disk` KWD https://github.com/Dasharo/open-source-firmware-validation/blob/uefi-sb-release/lib/ansible.robot#L18

Commit:
ansible: secure-boot: set sbctl checkout to 0.13 tag

Signed-off-by: Tomasz Żyjewski <tomasz.zyjewski@3mdeb.com>

TomaszAIR commented on ansible-roles/roles/secure-boot/tasks/common.yml:
```diff
@@ -0,0 +1,37 @@
+  tasks:
+    - name: Fetch sbctl repository
+      git:
+        repo: https://github.com/Foxboron/sbctl
```
fixed https://github.com/Dasharo/open-source-firmware-validation/pull/187/commits/9fb3c31e05d1c782edef0d1bce649ecf58e5e373

TomaszAIR commented on keywords.robot:
```diff
@@ -43,7 +42,7 @@ Login To Linux
         # this should not be a huge problem.
         # Read From Terminal Until    login:
-        Set Test Variable    ${DUT_CONNECTION_METHOD}    SSH
+        Set Global Variable    ${DUT_CONNECTION_METHOD}    SSH
```
In commit msg which introduced this change I added commit hash which changed this to `Set Test Variable` with a link to the documentation. Might help to find out why this was changed https://github.com/Dasharo/open-source-firmware-validation/commit/14519ee0ffa8ce0dc660892463983858cb269391

Commit:
lib: ansible: use Execute Reboot Command KWD

Signed-off-by: Tomasz Żyjewski <tomasz.zyjewski@3mdeb.com>

TomaszAIR commented on lib/ansible.robot:
```diff
@@ -0,0 +1,35 @@
+                Should Be Equal As Integers    ${rc}    0
+                Should Not Contain    ${output}    FAIL
+                Switch To Root User
+                Write Into Terminal    reboot
```
fixed https://github.com/Dasharo/open-source-firmware-validation/pull/187/commits/8cd7c3c0d648143e93d891b11bd78fdcbaf299b1

Commit:
lib: ansible: change msg when ANSIBLE_CONFIG not supported

Signed-off-by: Tomasz Żyjewski <tomasz.zyjewski@3mdeb.com>

TomaszAIR commented on lib/ansible.robot:
```diff
@@ -0,0 +1,35 @@
+                Read From Terminal Until    System Reboot
+            END
+        ELSE
+            Log    This keyword works only on QEMU...    INFO
```
Log message changed in https://github.com/Dasharo/open-source-firmware-validation/pull/187/commits/c8fdb295a2851f5c1b0c6a08e3e4c3f8b699d00e

I think that idea here was to set ANSIBLE_SUPPORT on those platforms that we at least once actually tested.

TomaszAIR commented on lib/bios/menus-ami.robot:
```diff
@@ -0,0 +1,156 @@
+*** Settings ***
+Documentation       Collection of keywords related to AMI menus
```
Here is exact AMI BIOS menu documentation that we tested https://cloud.3mdeb.com/index.php/s/oaLqc2D3B66YWF8 (it is also available publicly on https://www.mpl.ch/, logging in is needed). How we should store information about tested fw? Some variables in one of the robot files?

TomaszAIR commented on lib/bios/menus-dasharo.robot:
```diff
@@ -559,16 +376,24 @@ Save Changes
+    ...    ${main_menu_steps_to_reset} means how many times should
+    ...    arrow down be pressed to get to the Reset option in main
+    ...    settings menu
+    # robocop: disable=unused-argument
```
@m-iwanicki can you help there? I think you added this change (it was in my commit by rebasing and squashing some of them).

m-iwanicki replied:
@TomaszAIR @macpijan I think I changed how reset was done so it could be done without knowing/remembering how to get to Reset option from where we currently are.
I left those default unused arguments because there are other places where this keyword is used with them and if I remember correctly there were other PR's that used them that would stop working when merged.

TomaszAIR commented on lib/secure-boot-lib-ami.robot:
```diff
@@ -0,0 +1,228 @@
+    [Documentation]    This keyword enters Secure Boot menu after the platform was powered on.
+    Enter Setup Menu
+    Read From Terminal Until    ---/
+    Press Key N Times    1    ${ARROW_RIGHT}
```
@m-iwanicki can you help there? I think you added this change (it was in my commit by rebasing and squashing some of them).

Commit:
lib: wrap to 80 characters

Signed-off-by: Tomasz Żyjewski <tomasz.zyjewski@3mdeb.com>

TomaszAIR commented on lib/secure-boot-lib-ami.robot:
```diff
@@ -0,0 +1,228 @@
+    Press Enter
+
+Enter Secure Boot Menu And Return Construction
+    [Documentation]    This keyword enters Secure Boot menu after the platform was powered on. Returns Secure Boot menu construction.
```
Sorry, done here https://github.com/Dasharo/open-source-firmware-validation/pull/187/commits/6e8ca9f357895d808bba52012c2b36c42eb70906

TomaszAIR commented on lib/secure-boot-lib-common.py:
```diff
@@ -0,0 +1,27 @@
+    modified_content = ""
+    for line in lines:
+        if line.startswith("SHA1 Fingerprint"):
+            # Extract the fingerprint part, change it to uppercase and skip adding this line to modified content
```
Done https://github.com/Dasharo/open-source-firmware-validation/pull/187/commits/6e8ca9f357895d808bba52012c2b36c42eb70906#diff-7a7a5f018cf90c0995d0c9a7603d36eb90df8aed024497a4d408eab1b1c9bdd3R14

TomaszAIR commented on lib/secure-boot-lib-ami.robot:
```diff
@@ -0,0 +1,228 @@
+    [Documentation]    Enters (Advanced) Key Management menu and returns constructions.
+    ...    Should be called from secure boot menu
+    # robocop: disable=unused-argument
+    [Arguments]    ${sb_menu}=${EMPTY}
```
@m-iwanicki please correctly me if I am wrong, but I think that we wanted to use the same name of this KWD for Dasharo (https://github.com/Dasharo/open-source-firmware-validation/blob/uefi-sb-release/lib/secure-boot-lib-dasharo.robot#L68) and AMI (https://github.com/Dasharo/open-source-firmware-validation/blob/uefi-sb-release/lib/secure-boot-lib-ami.robot#L46) while they work differently. Plus sometimes this keyword is triggered with argument (https://github.com/Dasharo/open-source-firmware-validation/blob/uefi-sb-release/dasharo-security/secure-boot.robot#L490) and other times is not (https://github.com/Dasharo/open-source-firmware-validation/blob/uefi-sb-release/dasharo-security/secure-boot.robot#L326). Maybe it could be possible to get rid off this argument at all. @m-iwanicki WDYT?

m-iwanicki replied:
@TomaszAIR Yes in some tests this keyword is used with arguments (on Dasharo) but it's not needed on AMI BIOS.
When this keyword is used with argument then that means before using it there was used keyword that returned menu construction. Either we would need to change that or I guess we could always `Reenter Menu` to make sure we can read it from Terminal.

Commit:
lib: secure-boot-lib-ami: remove commented line

Signed-off-by: Tomasz Żyjewski <tomasz.zyjewski@3mdeb.com>

TomaszAIR commented on lib/secure-boot-lib-ami.robot:
```diff
@@ -0,0 +1,228 @@
+    ...    If disk is empty then chooses first disk in menu. At the end returns
+    ...    to the top of Boot menu
+    [Arguments]    ${file}    ${boot_option}    ${volume}
+    Press Key N Times    2    ${ARROW_LEFT}
```
Two times left is here to go into `Boot` tab in BIOS menu. Layout presented on page 60 of https://cloud.3mdeb.com/index.php/s/oaLqc2D3B66YWF8 so adding more drives/bootable entries does not influence that

TomaszAIR commented on lib/secure-boot-lib-ami.robot:
```diff
@@ -0,0 +1,228 @@
+
+Boot Efi File And Return Response
+    [Documentation]    Boots EFI file and returns response.
+    ...    If read_only_first_line is true then reads only up to \r\n else
```
@m-iwanicki could you also help here? Do you have any idea how this could be implemented better?

m-iwanicki replied:
@TomaszAIR @macpijan When booting hello_world system prints for very short time "Hello World" after whih it returns to Setup Menu. telnet reads for a very short time: "Hello World\r\n" and then reads Setup Menu which either clears screen or just overwrites this Hello World.
End result is that in output text from Read From Terminal there is no "Hello World" line.
This keyword is also used when trying to boot unsigned EFI file in which case it will return Error Dialog (that ends in `---/`.

Commit:
lib: secure-boot-lib-common: remove commented code

Signed-off-by: Tomasz Żyjewski <tomasz.zyjewski@3mdeb.com>

TomaszAIR commented on lib/secure-boot-lib-common.py:
```diff
@@ -0,0 +1,27 @@
+        if line.startswith("SHA1 Fingerprint"):
+            # Extract the fingerprint part, change it to uppercase and skip adding this line to modified content
+            continue
+            # fingerprint = line.replace("=", ":").upper()
```
Commented code removed https://github.com/Dasharo/open-source-firmware-validation/pull/187/commits/9e26f7fbfbe29009c4be571910c24744e25028db - also I changed there added comment to this function

Commit:
lib: secure-boot-lib-common: update Documentation section for KWD

Signed-off-by: Tomasz Żyjewski <tomasz.zyjewski@3mdeb.com>

TomaszAIR commented on lib/secure-boot-lib-common.robot:
```diff
@@ -0,0 +1,122 @@
+
+*** Keywords ***
+Check If Certificate Images For Tests Exists
+    [Documentation]    This keyword generates all the necessary images
```
Leftover, fixed https://github.com/Dasharo/open-source-firmware-validation/pull/187/commits/d2b5deb47d9d0bd7ed710ae95afc911a7e1420b0

TomaszAIR commented on lib/secure-boot-lib-common.robot:
```diff
@@ -0,0 +1,122 @@
+    Save Changes And Reset    3    5
+
+    # 2. Boot to Automatic provisioning tool
+    Boot Dasharo Tools Suite    USB    ${TRUE}
```
Explained in commit msg that adds this https://github.com/Dasharo/open-source-firmware-validation/commit/592683c7e4796a705eb5a45df09393e0c6cf098b

To perform this automatic provisioning we need to boot this special DTS image.

Commit:
lib: secure-boot-lib-common: use Execute Command In Terminal

Signed-off-by: Tomasz Żyjewski <tomasz.zyjewski@3mdeb.com>

Commit:
lib: usb-fid-msc-lib: add informative log for skipping KWD

Signed-off-by: Tomasz Żyjewski <tomasz.zyjewski@3mdeb.com>

Commit:
scripts: secure-boot: remove one level of dirs

Signed-off-by: Tomasz Żyjewski <tomasz.zyjewski@3mdeb.com>

Commit:
variables: remove not used vars for MPL-PIP4

Signed-off-by: Tomasz Żyjewski <tomasz.zyjewski@3mdeb.com>

TomaszAIR commented:
@macpijan I responded in all threads. I think that @m-iwanicki @PLangowski @DaniilKl might help if there will be additional questions.
